typedb/api/concept/value/value.py

```diff
@@ -18,105 +18,115 @@
-    def is_boolean(self):
-        return False
-
-    def is_long(self):
-        return False
-
-    def is_double(self):
-        return False
-
-    def is_string(self):
-        return False
-
-    def is_datetime(self):
-        return False
-
-    def as_remote(self, transaction: "TypeDBTransaction"):
-        raise TypeDBClientException.of(VALUE_HAS_NO_REMOTE)
-
-    def to_json(self) -> Mapping[str, Union[str, int, float, bool]]:
-        return {
-            "value_type": str(self.get_value_type()),
-            "value": self.get_value(),
-        }
-
-
-class BooleanValue(Value, ABC):
+    def as_value(self) -> Value:
+        return self
 
     def is_boolean(self) -> bool:
-        return True
-
-    @abstractmethod
-    def get_value(self) -> bool:
         pass
 
-
-class LongValue(Value, ABC):
-
     def is_long(self) -> bool:
-        return True
-
-    @abstractmethod
-    def get_value(self) -> int:
         pass
 
-
-class DoubleValue(Value, ABC):
-
     def is_double(self) -> bool:
-        return True
-
-    @abstractmethod
-    def get_value(self) -> float:
         pass
 
-
-class StringValue(Value, ABC):
-
     def is_string(self) -> bool:
-        return True
+        pass
 
-    @abstractmethod
-    def get_value(self) -> str:
+    def is_datetime(self) -> bool:
         pass
 
+    def as_boolean(self) -> bool:
+        pass
 
-class DateTimeValue(Value, ABC):
+    def as_long(self) -> int:
+        pass
 
-    def is_datetime(self) -> bool:
-        return True
+    def as_double(self) -> float:
+        pass
 
-    @abstractmethod
-    def get_value(self) -> datetime:
+    def as_string(self) -> str:
+        pass
+
+    def as_datetime(self) -> datetime:
         pass
 
     def to_json(self) -> Mapping[str, Union[str, int, float, bool]]:
         return {
             "value_type": str(self.get_value_type()),
-            "value": self.get_value().isoformat(timespec='milliseconds')
+            "value": self.get_value() if not self.is_datetime() else self.get_value().isoformat(timespec='milliseconds')
         }
+
+
+class _ValueType:
+
+    def __init__(self, is_writable: bool, is_keyable: bool, native_object):
+        self._is_writable = is_writable
+        self._is_keyable = is_keyable
+        self._native_object = native_object
+
+    def is_writable(self) -> bool:
+        return self._is_writable
+
+    def is_keyable(self) -> bool:
+        return self._is_keyable
+
+    @property
+    def native_object(self):
+        return self._native_object
+
+
+class ValueType(Enum):
+    OBJECT = _ValueType(False, False, Object)
+    BOOLEAN = _ValueType(True, False, Boolean)
+    LONG = _ValueType(True, True, Long)
+    DOUBLE = _ValueType(True, False, Double)
+    STRING = _ValueType(True, True, String)
+    DATETIME = _ValueType(True, True, DateTime)
+
+    @property
+    def native_object(self):
+        return self.value.native_object
+
+    def __str__(self):
+        mapping = {ValueType.OBJECT: "object",
```

Comment:
We can't use loop here, because we need to specify a string representation for each value. But if it is always a lowercase enum variant, we don't need neither dict, nor loop.